### ISSUE TITLE
[FEATURE REQUEST] Realtime Lighting Preview in Editor

### ISSUE THREAD

**Issue by TarekLP** [Duude92/hammertime#18]
**Is your feature request related to a problem? Please describe.**
This Feature Request is related to a problem which I have with other editors, it is Realtime lighting like in Hammer++

**Describe the solution you'd like**
Adding Realtime Lighting

**Describe alternatives you've considered**
Using Hammer++ for a GameDev Project which uses Source BSPs, but I want to use Hammertime because I want to support this Project and not Hammer++
**Additional context**
Nothing Else

**Comment by Duude92** [comment]
It's been a while, since I've started working on that feature with no updates.
But work is going, so I'll share some results
So, initially I tried to implement realtime shadowmapping utilizing GPU resources, and turns out - that was terrible idea.
Shadowmapping is good idea for dynamic scenes with camera stick to the ground/scene itself, the issue was in rendering shadows over large scene on far distance (it could be fixed with array of shadow maps, but, i'm not into it)
<details>
<img width="1280" height="669" alt="Image" src="https://github.com/user-attachments/assets/bc547145-8dd1-4936-b5c5-bd0682e8327a" />

<img width="1280" height="672" alt="Image" src="https://github.com/user-attachments/assets/510d0d86-fbf0-4bbc-89d0-87910d729146" />
</details>
So later (yeah, a lot of later), I decided to replace shadowmapping to static baked lightmaps (yes, lightmaps, but it only bakes shadows for now).
Yes, it's slower, so I've used parallelism to divide each lightmap baking to free cpu thread, and applied BVH tree search to avoid testing each surface with each other, so on my 12 thread cpu dust2 full bake takes 30-40seconds to perform.
And, since it still has some artifacts, I would continue to fix them, but for now it looks like this, and 
[shadowmap](https://github.com/Duude92/hammertime/tree/shadowmap)
 branch has all of this.
Here is couple screenshots of current stage of feature development:
<details>

<img width="1572" height="756" alt="Image" src="https://github.com/user-attachments/assets/70443058-ac7c-417b-a7db-5817981532b4" />

<img width="1570" height="755" alt="Image" src="https://github.com/user-attachments/assets/9eb0f44f-c083-4dd2-b382-d49b3f598d5f" />

<img width="1571" height="759" alt="Image" src="https://github.com/user-attachments/assets/3c3aa14c-ee6a-41c7-ac48-b8f7eb317fd3" />

<img width="1569" height="759" alt="Image" src="https://github.com/user-attachments/assets/54578e15-9e7f-4580-ae2f-6e3ef63bd858" />

</details>

**Comment by TarekLP** [comment]
This is looking better and better

**Comment by TarekLP** [comment]
This is looking better and better

**Comment by Duude92** [comment]
This was automatically opened, but, I want to keep it open for now, since it has some issues to discuss
For now, feature is merged into main build, and now it is possible to use it.
How to use it (for now, as I still hesitate on where it should be placed and how to be used):
<details>
<img width="1920" height="892" alt="Image" src="https://github.com/user-attachments/assets/f37119f1-b704-4f44-baf1-b7f0f16edbb2" />
</details>
Red marks lead to shadowmap baking function, and green one is to enable/disable displaying shadow.
Shadow settings are empty for now, as I not quite sure what to enable here

**Comment by Cianez** [comment]
> If I'm correct, light_environment is the requirement for hlvis to build environment lighting Also please note that this feature is applicable for environment light only, baking entities such as light_spot/light, e.t.c. is not implemented yet

Yeah, I suspected that. Though, are you planning on adding the feature on all light type entities in the future?

**Comment by Duude92** [comment]
I don't want to
But I have to
Probably after minimal source engine environment implementation would be done

**Comment by Duude92** [comment]
> Hey, how do I activate the lighting preview?

First, you have to setup your scene (add env_environment)
Second, you have camera tool, on its right sidebar panel, there button to bake light
Third, on top panel, there is icon with sun, this will enable/disable baked light displaying

**Comment by itsthereeps** [comment]
Somehow was able to get lighting previews to actually bake however it dosent actually show the lighting and just looks wrong

<img width="2137" height="1274" alt="Image" src="https://github.com/user-attachments/assets/e8ca974e-65fb-4523-a7bf-766e4ae36387" />

<img width="1919" height="990" alt="Image" src="https://github.com/user-attachments/assets/1f4c08ac-beac-4406-9a62-0759c8ea98e0" />
 

<img width="2560" height="1440" alt="Image" src="https://github.com/user-attachments/assets/ce92789c-5288-4780-9422-460cae595f89" />

**Comment by itsthereeps** [comment]
also it for some reason casts the light on outside faces
